### PR TITLE
[Technical] Fixed incorrect culture comparison

### PR DESCRIPTION
**Pull request by duncanpMS**
Fixed bugs in assembly resolver

### CODE REVIEW

File: Common/AssemblyResolver.cs [hunk 1]
```diff
@@ -30,10 +33,12 @@ public AssemblyResolver(ILogger logger, params string[] rootSearchPaths)
             if (logger == null)
             {
                 throw new ArgumentNullException("logger");
-            } else if (rootSearchPaths == null || rootSearchPaths.Length < 1)
```
Comment (duncanpMS) [inline Common/AssemblyResolver.cs]: We don't normally include the "else" when checking arguments.

File: Common/AssemblyResolver.cs [hunk 2]
```diff
@@ -45,13 +50,18 @@ public AssemblyResolver(ILogger logger, params string[] rootSearchPaths)
             AppDomain.CurrentDomain.AssemblyResolve += CurrentDomain_AssemblyResolve;
         }
 
-        public Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args) // set to public for test purposes
+        private Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
```
Comment (duncanpMS) [inline Common/AssemblyResolver.cs]: Changed the tests so this does not need to be public

File: Common/AssemblyResolver.cs [hunk 3]
```diff
@@ -59,7 +69,13 @@ public AssemblyResolver(ILogger logger, params string[] rootSearchPaths)
                 {
                     asm = Assembly.LoadFile(file);
 
-                    if (string.Equals(args.Name, asm.FullName, StringComparison.CurrentCultureIgnoreCase))
```
Comment (duncanpMS) [inline Common/AssemblyResolver.cs]: Should using OrdinalIgnoreCase when comparing file names: https://msdn.microsoft.com/en-us/library/dd465121(v=vs.110).aspx

File: Common/AssemblyResolver.cs [hunk 4]
```diff
@@ -59,7 +69,13 @@ public AssemblyResolver(ILogger logger, params string[] rootSearchPaths)
...
+                    if (
+                        // If the input was e.g foo.dll then compare against the file name...
+                        (isFileName && string.Equals(Path.GetFileName(asm.Location), fileName,  StringComparison.OrdinalIgnoreCase))
```
Comment (duncanpMS) [inline Common/AssemblyResolver.cs]: The previous code didn't work correctly if the supplied argument was a file name.

File: Common/AssemblyResolver.cs [hunk 5]
```diff
@@ -75,28 +91,28 @@ public AssemblyResolver(ILogger logger, params string[] rootSearchPaths)
...
                 return input;
             }
 
-            string result = input;
-            if (input.Contains(" "))
-            {
-                // If the assembly name has multiple words (seperated by spaces), use only the first word
-                // (e.g. "foo bar" -> "foo")
-                string[] parts = input.Split(new char[] { ' ' });
-                result = parts[0].Substring(0, parts[0].Length - 1);
```
Comment (duncanpMS) [inline Common/AssemblyResolver.cs]: This code was left over from the initial prototype. It doesn't correctly parse all valid full assembly names (e.g. those containing spaces). I've added a test case for this too.

File: PluginGenerator/JarBuilder.cs [hunk 6]
```diff
@@ -170,7 +170,7 @@ private void WriteManifestFile(string manifestFilePath)
             StringBuilder sb = new StringBuilder();
             foreach (KeyValuePair<string, string> kvp in this.manifestProperties)
             {
-                sb.AppendFormat(System.Globalization.CultureInfo.CurrentCulture,
```
Comment (duncanpMS) [inline PluginGenerator/JarBuilder.cs]: We don't want culture-sensitive behaviour when building the key-value pair strings, or when building the command line arguments below.

File: Tests/CommonTests/AssemblyResolverTests.cs [hunk 7]
```diff
@@ -20,145 +20,244 @@ public class AssemblyResolverTests
     {
         public TestContext TestContext { get; set; }
 
+        #region Tests
+
         /// <summary>
-        /// Compiles the supplied code into a new assembly
+        /// Tests the loading of an assembly with a single type and no dependencies. This should succeed even without the AssemblyResolver.
         /// </summary>
-        private static Assembly CompileAssembly(string code, string outputFilePath, ILogger logger)
```
Comment (duncanpMS) [inline Tests/CommonTests/AssemblyResolverTests.cs]: I moved the helper methods to the end of the file, as usual in our tests.

File: Tests/CommonTests/AssemblyResolverTests.cs [hunk 8]
```diff
@@ -20,145 +20,244 @@ public class AssemblyResolverTests
...
                 }
-                Assert.Fail("Test setup error: failed to create dynamic assembly. See the test output for compiler output");
-            }
 
-            return result.CompiledAssembly;
-        }
+                // Assert
+                Assert.IsNotNull(simpleObject);
+                Assert.AreEqual<string>("SimpleProgram", simpleObject.GetType().ToString());
+                AssertResolverCaller(resolver);
```
Comment (duncanpMS) [inline Tests/CommonTests/AssemblyResolverTests.cs]: Added a check that the resolver was actually used.

File: Tests/CommonTests/AssemblyResolverTests.cs [hunk 9]
```diff
@@ -20,145 +20,244 @@ public class AssemblyResolverTests
...
             // Arrange
-            TestLogger logger = new TestLogger();
             string testFolder = TestUtils.CreateTestDirectory(this.TestContext);
-            string simpleAssemblyPath = Path.Combine(testFolder, "SimpleAssembly.dll");
-            Assembly simpleAssembly = CompileSimpleAssembly(simpleAssemblyPath, logger);
+            Assembly testAssembly = CompileSimpleAssembly("SimpleAssemblyByFullName.dll", testFolder, new TestLogger());
 
-            object simpleObject = null;
+            // Act
+            Assembly resolvedAssembly = AssertAssemblyLoadSucceedsOnlyWithResolver("SimpleAssemblyByFullName, Version = 0.0.0.0, Culture = neutral, PublicKeyToken = null", testFolder);
```
Comment (duncanpMS) [inline Tests/CommonTests/AssemblyResolverTests.cs]: Test loading using the assembly name format i.e. "name, version=xxx, culture=xxx..." etc

File: Tests/CommonTests/AssemblyResolverTests.cs [hunk 10]
```diff
@@ -20,145 +20,244 @@ public class AssemblyResolverTests
...
-                    foreach (Type type in assembly.GetExportedTypes())
-                    {
-                        if (!type.IsAbstract)
-                        {
-                            simpleObject = Activator.CreateInstance(type);
-                        }
-                    }
-                }
-            }
+            Assembly resolvedAssembly = AssertAssemblyLoadSucceedsOnlyWithResolver("SimpleAssemblyByFileName.dll", testFolder);
```
Comment (duncanpMS) [inline Tests/CommonTests/AssemblyResolverTests.cs]: Test using file name.

File: Tests/CommonTests/AssemblyResolverTests.cs [hunk 11]
```diff
@@ -20,145 +20,244 @@ public class AssemblyResolverTests
...
+            AssertExpectedAssemblyLoaded(testAssembly, resolvedAssembly);
         }
 
         /// <summary>
-        /// Tests the case where assembly resolution should fail correctly.
+        /// Tests the case where assembly resolution should succeed.
         /// </summary>
         [TestMethod]
-        public void TestAssemblyResolutionFail()
+        public void AssemblyResolver_ResolutionByFullAssemblyNameWithSpace_Succeeds()
```
Comment (duncanpMS) [inline Tests/CommonTests/AssemblyResolverTests.cs]: ...and tests for full names and file names containing spaces.

File: Tests/CommonTests/AssemblyResolverTests.cs [hunk 12]
```diff
@@ -20,145 +20,244 @@ public class AssemblyResolverTests
...
+
+            // Assert
+            AssertExpectedAssemblyLoaded(testAssembly, resolvedAssembly);
+        }
+
+        /// <summary>
+        /// Tests the case where assembly resolution should succeed.
+        /// </summary>
+        [TestMethod]
+        public void AssemblyResolver_VersionAssemblyRequested()
```
Comment (duncanpMS) [inline Tests/CommonTests/AssemblyResolverTests.cs]: New test to check version is taken into account.

File: Tests/CommonTests/AssemblyResolverTests.cs [hunk 13]
```diff
@@ -20,145 +20,244 @@ public class AssemblyResolverTests
...
             // Act
             Assembly resolveResult;
-            using (AssemblyResolver assemblyResolver = new AssemblyResolver(logger, testFolder))
+
+            // Create a test logger that will only record output from the resolver
+            // so we can check it has been called
+            using (AssemblyResolver resolver = new AssemblyResolver(new TestLogger(), searchPath))
             {
-                ResolveEventArgs resolveEventArgs = new ResolveEventArgs(simpleAssembly.FullName, this.GetType().Assembly);
-                resolveResult = assemblyResolver.CurrentDomain_AssemblyResolve(this, resolveEventArgs);
```
Comment (duncanpMS) [inline Tests/CommonTests/AssemblyResolverTests.cs]: The previous tests expose a couple of methods publicly and called them from the tests. The new tests trigger a real "resolve" event instead.

File: Tests/CommonTests/AssemblyResolverTests.cs [hunk 14]
```diff
@@ -20,145 +20,244 @@ public class AssemblyResolverTests
...
+
+        private static void AssertAssemblyLoadFails(string asmRef)
+        {
+            AssertException.Expect<FileNotFoundException>(() => Assembly.Load(asmRef));
+        }
+
+        private Assembly AssertAssemblyLoadSucceedsOnlyWithResolver(string asmRef, string searchPath)
+        {
+            // Check the assembly load fails without the assembly resolver
+            AssertAssemblyLoadFails(asmRef);
```
Comment (duncanpMS) [inline Tests/CommonTests/AssemblyResolverTests.cs]: Checks that resolution will fail without the resolver - means we know that if resolution succeeds it is due to the resolver.

File: Tests/CommonTests/AssemblyResolverTests.cs [hunk 15]
```diff
@@ -20,145 +20,244 @@ public class AssemblyResolverTests
...
+
             // Act
             Assembly resolveResult;
-            using (AssemblyResolver assemblyResolver = new AssemblyResolver(logger, testFolder))
+
+            // Create a test logger that will only record output from the resolver
+            // so we can check it has been called
```
Comment (duncanpMS) [inline Tests/CommonTests/AssemblyResolverTests.cs]: This comment is no longer correct: I added the "ResolverCalled" property instead as it is more reliable and cleaner than checking for log output.

File: Common/AssemblyResolver.cs [hunk 16]
```diff
@@ -75,28 +91,28 @@ public AssemblyResolver(ILogger logger, params string[] rootSearchPaths)
...
-            }
+            AssemblyName assemblyName = new AssemblyName(input);
+            return assemblyName.Name + DllExtension;
+        }
 
-            return  result + ".dll";
+        private static bool IsFileName(string fileName)
```
Comment (bgavrilMS) [inline Common/AssemblyResolver.cs]: IsAssemblyByFileName ? Since there are other areas where this logic happens, consider extracting to an Utils class.

**Comment by bgavrilMS** [comment]
LGTM

**Pull request merged**